In [1]:
#Importing necessary libraries
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt

In [2]:
#changing the work directory
os.chdir('/Users/admin/Desktop/Synergyx Dataset')

In [3]:
#reading the data
data = pd.read_csv('Synergix_data_preprocessed_new.csv')
data.head()

Page_traffic  Unit_price  Units_sold    Segment  1_Star_Rating  \
0        5835.0   22.214389        2071   Skincare           99.0   
1        1881.0   11.870778         681   Skincare          118.0   
2        2477.0   27.645714         875     Makeup          108.0   
3        4087.0   16.896848        1396     Makeup           82.0   
4        1446.0   16.357664         822  Hair Care           29.0   

   2_Star_Rating  3_Star_Rating  4_Star_Rating  5_Star_Rating  Image_Count  \
0           72.0          194.0          453.0         3272.0          6.0   
1           73.0          200.0          430.0         3084.0          9.0   
2           85.0          232.0          511.0         3402.0         10.0   
3           64.0          167.0          303.0         2058.0          5.0   
4           24.0           88.0          200.0         1221.0          0.0   

   ...  Sum of Spend  Online_Clicks  Online_Cost  Online_Impressions  \
0  ...           0.0         1271.0  3102.448115            238075.0   
1  ...           0.0          494.0  1723.716359             75868.0   
2  ...           0.0            0.0     0.000000                 0.0   
3  ...           0.0            0.0     0.000000                 0.0   
4  ...           0.0            0.0     0.000000                 0.0   

   num_unique_campaigns  Clicks_video  Cost_video  Impressions_video  \
0                   6.0        3479.0    40244.28          3856872.0   
1                   2.0           0.0        0.00                0.0   
2                   0.0           0.0        0.00                0.0   
3                   0.0           0.0        0.00                0.0   
4                   0.0        6024.0    77040.81          7539723.0   

   num_unique_campaigns_offline  Units_sold>1000  
0                           2.0                1  
1                           0.0                0  
2                           0.0                0  
3                           0.0                1  
4                           2.0                0  

[5 rows x 26 columns]

In [4]:
#drop Units_sold column
data.drop(columns = 'Units_sold', axis = 1, inplace = True)

In [5]:
#check the data
data.head()

Page_traffic  Unit_price    Segment  1_Star_Rating  2_Star_Rating  \
0        5835.0   22.214389   Skincare           99.0           72.0   
1        1881.0   11.870778   Skincare          118.0           73.0   
2        2477.0   27.645714     Makeup          108.0           85.0   
3        4087.0   16.896848     Makeup           82.0           64.0   
4        1446.0   16.357664  Hair Care           29.0           24.0   

   3_Star_Rating  4_Star_Rating  5_Star_Rating  Image_Count  Bullet_Count  \
0          194.0          453.0         3272.0          6.0           6.0   
1          200.0          430.0         3084.0          9.0           8.0   
2          232.0          511.0         3402.0         10.0           7.0   
3          167.0          303.0         2058.0          5.0           7.0   
4           88.0          200.0         1221.0          0.0           8.0   

   ...  Sum of Spend  Online_Clicks  Online_Cost  Online_Impressions  \
0  ...           0.0         1271.0  3102.448115            238075.0   
1  ...           0.0          494.0  1723.716359             75868.0   
2  ...           0.0            0.0     0.000000                 0.0   
3  ...           0.0            0.0     0.000000                 0.0   
4  ...           0.0            0.0     0.000000                 0.0   

   num_unique_campaigns  Clicks_video  Cost_video  Impressions_video  \
0                   6.0        3479.0    40244.28          3856872.0   
1                   2.0           0.0        0.00                0.0   
2                   0.0           0.0        0.00                0.0   
3                   0.0           0.0        0.00                0.0   
4                   0.0        6024.0    77040.81          7539723.0   

   num_unique_campaigns_offline  Units_sold>1000  
0                           2.0                1  
1                           0.0                0  
2                           0.0                0  
3                           0.0                1  
4                           2.0                0  

[5 rows x 25 columns]

In [6]:
#Storing the ratio in a list named Rating_ratio
Rating_ratio = []
for row in data.values:
    if(row[3]+row[4] == 0):
        if(row[6]+row[7] == 0):
            #If all the ratings are zero then overall rating ratio will also be zero
            Rating_ratio.append(0.0)
        else:
            #If only the numerator(1 and 2 star) ratings are zero then add -99999 to the list temporarily which 
            #will be taken care of in the next part of the code
            Rating_ratio.append(-99999)
    else:
        Rating_ratio.append((int(row[6])+int(row[7]))/(int(row[4])+int(row[3])));
              

#replacing -99999 with the maximum rating in the list
max_rating = max(Rating_ratio)
for x in range(len(Rating_ratio)):
    if(Rating_ratio[x] == -99999):
        Rating_ratio[x] = max_rating

In [7]:
#adding the column 'Good_By_Bad_Rating' to the dataframe
data['Good_By_Bad_Rating'] = Rating_ratio

In [8]:
#drop the rating columns
data.drop(columns = ['1_Star_Rating', '2_Star_Rating', '3_Star_Rating', '4_Star_Rating', '5_Star_Rating'], inplace = True)

In [9]:
#check the dataset
data.head().transpose()

0            1          2          3  \
Page_traffic                       5835.0       1881.0     2477.0     4087.0   
Unit_price                      22.214389    11.870778  27.645714  16.896848   
Segment                          Skincare     Skincare     Makeup     Makeup   
Image_Count                           6.0          9.0       10.0        5.0   
Bullet_Count                          6.0          8.0        7.0        7.0   
Title_Count                          89.0        126.0      152.0      117.0   
Description_Length                  141.0        149.0      186.0      209.0   
num_unique_Search_Terms               3.0         13.0        4.0        7.0   
Organic Search Rank                   3.0         16.0       20.0       20.0   
Num_of_Promotions                     0.0          0.0        0.0        0.0   
Sum of Spend                          0.0          0.0        0.0        0.0   
Online_Clicks                      1271.0        494.0        0.0        0.0   
Online_Cost                   3102.448115  1723.716359        0.0        0.0   
Online_Impressions               238075.0      75868.0        0.0        0.0   
num_unique_campaigns                  6.0          2.0        0.0        0.0   
Clicks_video                       3479.0          0.0        0.0        0.0   
Cost_video                       40244.28          0.0        0.0        0.0   
Impressions_video               3856872.0          0.0        0.0        0.0   
num_unique_campaigns_offline          2.0          0.0        0.0        0.0   
Units_sold>1000                         1            0          0          1   
Good_By_Bad_Rating              21.783626    18.397906  20.274611  16.171233   

                                      4  
Page_traffic                     1446.0  
Unit_price                    16.357664  
Segment                       Hair Care  
Image_Count                         0.0  
Bullet_Count                        8.0  
Title_Count                       165.0  
Description_Length                158.0  
num_unique_Search_Terms             4.0  
Organic Search Rank                16.0  
Num_of_Promotions                   0.0  
Sum of Spend                        0.0  
Online_Clicks                       0.0  
Online_Cost                         0.0  
Online_Impressions                  0.0  
num_unique_campaigns                0.0  
Clicks_video                     6024.0  
Cost_video                     77040.81  
Impressions_video             7539723.0  
num_unique_campaigns_offline        2.0  
Units_sold>1000                       0  
Good_By_Bad_Rating            26.811321

In [10]:
#get the category names
data['Segment'].unique()

array(['Skincare', 'Makeup', 'Hair Care'], dtype=object)

## One-hot Encoding

In [11]:
#one-hot encoding
data = pd.get_dummies(data, columns = ['Segment'], prefix = 'Segment', drop_first = True)
data.head()

Page_traffic  Unit_price  Image_Count  Bullet_Count  Title_Count  \
0        5835.0   22.214389          6.0           6.0         89.0   
1        1881.0   11.870778          9.0           8.0        126.0   
2        2477.0   27.645714         10.0           7.0        152.0   
3        4087.0   16.896848          5.0           7.0        117.0   
4        1446.0   16.357664          0.0           8.0        165.0   

   Description_Length  num_unique_Search_Terms  Organic Search Rank  \
0               141.0                      3.0                  3.0   
1               149.0                     13.0                 16.0   
2               186.0                      4.0                 20.0   
3               209.0                      7.0                 20.0   
4               158.0                      4.0                 16.0   

   Num_of_Promotions  Sum of Spend  ...  Online_Impressions  \
0                0.0           0.0  ...            238075.0   
1                0.0           0.0  ...             75868.0   
2                0.0           0.0  ...                 0.0   
3                0.0           0.0  ...                 0.0   
4                0.0           0.0  ...                 0.0   

   num_unique_campaigns  Clicks_video  Cost_video  Impressions_video  \
0                   6.0        3479.0    40244.28          3856872.0   
1                   2.0           0.0        0.00                0.0   
2                   0.0           0.0        0.00                0.0   
3                   0.0           0.0        0.00                0.0   
4                   0.0        6024.0    77040.81          7539723.0   

   num_unique_campaigns_offline  Units_sold>1000  Good_By_Bad_Rating  \
0                           2.0                1           21.783626   
1                           0.0                0           18.397906   
2                           0.0                0           20.274611   
3                           0.0                1           16.171233   
4                           2.0                0           26.811321   

   Segment_Makeup  Segment_Skincare  
0           False              True  
1           False              True  
2            True             False  
3            True             False  
4           False             False  

[5 rows x 22 columns]

In [12]:
#Separating the independent and target variables
y = data['Units_sold>1000']
X = data.drop(columns = ['Units_sold>1000'], axis = 1)

## Performing KFold cross validation

<p style = 'color:green'><b>Run all the cells above before you begin</b><p>

In [38]:
#import libraries
from sklearn.model_selection import KFold
from sklearn.model_selection import cross_validate
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import make_pipeline
from sklearn.neighbors import KNeighborsClassifier
#from sklearn.metrics import f1_score

In [39]:
#define cross-validation method to use
kf = KFold(n_splits=5, random_state=42, shuffle=True)

In [40]:
#setting the value k or neighbors at 10
knn = KNeighborsClassifier(10)

In [41]:
#creating a pipeline when used will first scale the data and then make predictions using knn
pipeline = make_pipeline(StandardScaler(), knn)

Now, let us perform cross validation. The cross_validate primarily takes in these areguments- the model you want to use, the dependent and the independent variables, the folds of cross validation and the evaluation metric, and if you also want the scores on the train set.

In our case the model is part of the pipeline instance. X and y are our independent and dependent variables respectively, the cross validation folds are saved in the kf instance and the evaluation metric is the f1 score. So let us implement this code.

In [34]:
# Perform k-fold cross-validation
cross_validate_results = cross_validate(pipeline, X, y, cv=kf, scoring='f1', return_train_score=True)

Lets understand what has happened when we ran this cell-
- In a loop of 5, Out of the 5 folds we created above 4 will function as training data and fifth fold as the test data
- First, the scaler.fit_transform will be applied to the training data which in our case are the 4 folds
- Then, scaler.transform will be applied on the test data which is the final fold.

After this knn model would run on the training dataset and store the f1 score of the model on the test and the train data.

And this will repeat 5 times, where in each iteration a different fold becomes the test data.

In [44]:
test_scores = cross_validate_results['test_score']
print("Cross-validation scores:", test_scores)
print("Cross-validation Mean score:", test_scores.mean())
print("Cross-validation STD:", test_scores.std())

Cross-validation scores: [0.85786802 0.85714286 0.8587071  0.85330444 0.85494277]
Cross-validation Mean score: 0.8563930367330199
Cross-validation STD: 0.0019867247057753


This tells us that this model has low variance. To look at the bias, we would need to look at the training scores. Lets do that in the code below.

In [45]:
cross_validate_results['train_score']

array([0.88608527, 0.88595885, 0.8854266 , 0.88668407, 0.88526159])

The training scores are around 88%. So wecan say that the remianing 12% represents error or the information which this model was not able to capture. This is the the bias of the model. This bias is neither too high nor too low. Additionally, the gap between the train and the test score is just 3% indicating limited overfitting.